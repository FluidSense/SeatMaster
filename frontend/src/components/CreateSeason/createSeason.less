.choose-date-button {
  margin-bottom: 1rem !important;
  margin-left: 0 !important;
}

#appSeason>div {
  display: grid;
  grid-template-columns: 50% 50%;
}

#appSeason {
  margin-bottom: 2rem;
}

#roomSeason>div {
  display: grid;
  grid-template-columns: 50% 50%;
}

#roomSeason {
  margin-bottom: 2rem;
}


@media only screen and (max-width: 600px) {

  #appSeason {
    margin-bottom: 0;
  }

  #appSeason>div,
  #roomSeason>div {
    display: flex;
    flex-direction: column;
  }

  .choose-date-button {
    width: 100%;
  }

  #roomSeason {
    margin-bottom: 1rem;
  }

  #new-season-btn {
    width: 100%;
  }
}
