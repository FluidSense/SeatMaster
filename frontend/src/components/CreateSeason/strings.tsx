export const _ROOM_START = 'Room period opens';
export const _ROOM_END = 'Room period ends';
export const _SEASON_START = 'Application season opens';
export const _SEASON_END = 'Application season ends';
export const _CREATE_NEW_SEASON = 'Create new application season';
export const _ROOM_END_TOO_EARLY = 'Room period ends before it starts';
export const _SEASON_END_TOO_EARLY = 'The application season ends before it starts';
export const _NEW_APPLICATION_SEASON = 'New Application Season';
export const _UPDATE_APPLICATION_SEASON = 'Update Application Season';
export const _ERROR_MESSAGE = 'Could not connect to the server';
export const _UPDATE_CURRENT_SEASON = 'Update current season';
export const _EXISTING_CURRENT_SEASON = 'There already exists a season';
