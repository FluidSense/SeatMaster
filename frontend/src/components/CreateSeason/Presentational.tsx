import { Hovedknapp } from 'nav-frontend-knapper';
import { SkjemaGruppe } from 'nav-frontend-skjema';
import { SkjemaelementFeil } from 'nav-frontend-skjema/lib/skjemaelement-feilmelding';
import React from 'react';
import { _CREATE_NEW_SEASON, _NEW_APPLICATION_SEASON } from './strings';

interface IProps {
  alertPeriodEndBeforeStart?: SkjemaelementFeil;
  alertApplicationEndBeforeStart?: SkjemaelementFeil;
  createFields: (index: number, end: number) => JSX.Element[];
  buttonDisable: boolean;
  postApplicationSeason: () => void;
  alertFail?: JSX.Element;
}

const Presentational: React.FunctionComponent<IProps> = (props) => {
  const {
    alertApplicationEndBeforeStart,
    alertPeriodEndBeforeStart,
    createFields,
    buttonDisable,
    postApplicationSeason,
    alertFail,
  } = props;
  return (
    <>
      <div id="new-application-season">
        <h1>{_NEW_APPLICATION_SEASON}</h1>
        <SkjemaGruppe feil={alertPeriodEndBeforeStart}> {createFields(0, 2)} </SkjemaGruppe>
        <SkjemaGruppe feil={alertApplicationEndBeforeStart}> {createFields(2, 4)} </SkjemaGruppe>
        <Hovedknapp id={'new-season-btn'} disabled={buttonDisable} onClick={postApplicationSeason}>
          {_CREATE_NEW_SEASON}
        </Hovedknapp>
        {alertFail}
      </div>
    </>
  );
};

export default Presentational;
