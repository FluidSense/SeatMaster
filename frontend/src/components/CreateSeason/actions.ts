import { AnyAction, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { IPostApplicationSeason } from '../../API/interfaces';
import { UPDATE_APPLICATION_SEASON } from '../ApplicationSeason/constants';
import { postSeason, putSeason } from './../../API/calls';
import { SUBMITTED_APPLICATION_SEASON } from './strings';

const newSeasonCreated = (payload: boolean) => ({
  payload,
  type: SUBMITTED_APPLICATION_SEASON,
});

const updateSeason = (payload: boolean) => ({
  payload,
  type: UPDATE_APPLICATION_SEASON,
});

export const postNewSeason = (body: IPostApplicationSeason):
  ThunkAction<Promise<void>, {}, {}, AnyAction> => async (dispatch: Dispatch) => {
    const result = await postSeason(body);
    if (result) dispatch(newSeasonCreated(true));
    else dispatch(newSeasonCreated(false));
  };

export const putNewSeason = (body: IPostApplicationSeason, id: number):
  ThunkAction<Promise<void>, {}, {}, AnyAction> => async (dispatch: Dispatch) => {
    const result = await putSeason(body, id);
    if (result) dispatch(updateSeason(true));
    else dispatch(updateSeason(false));
  };
