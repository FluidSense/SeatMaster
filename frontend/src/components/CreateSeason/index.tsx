import moment, { Moment } from 'moment';
import AlertStripe from 'nav-frontend-alertstriper';
import React, { Component } from 'react';
import { Redirect } from 'react-router';
import { postSeason } from '../../API/calls';
import DateInputField from '../DateInputField';
import Presentational from './Presentational';
import {
  _ERROR_MESSAGE,
  _ROOM_END,
  _ROOM_END_TOO_EARLY,
  _ROOM_START,
  _SEASON_END,
  _SEASON_END_TOO_EARLY,
  _SEASON_START,
} from './strings';

export interface IState {
  periodStart: Moment;
  periodEnd: Moment;
  roomStart: Moment;
  roomEnd: Moment;
  redirect: boolean;
  showAlert: boolean;
  [key: string]: Moment | boolean;
}

// Need to be in this order to match the proper fields
const inputTextArray = [
  _SEASON_START,
  _SEASON_END,
  _ROOM_START,
  _ROOM_END,
];

const errorObjectSeasonEndTooEarly = { feilmelding: _SEASON_END_TOO_EARLY };
const errorObjectRoomEndTooEarly = { feilmelding: _ROOM_END_TOO_EARLY };

export const setTime = (day: Moment) => {
  const newDay = moment(day);
  newDay.set({
    hour: 23,
    minute: 59,
  });
  return newDay;
};

// Format to match backend model
export const format = 'YYYY-MM-DD HH:mm:ss.SSS';

// tslint:disable-next-line:class-name
class CreateSeason extends Component<{}, IState> {
  constructor(props: object) {
    super(props);
    const currentTime = setTime(moment());
    const nextMonth = setTime(moment().add(1, 'month'));
    this.state = {
      periodEnd: nextMonth,
      periodStart: currentTime,
      redirect: false,
      roomEnd: nextMonth,
      roomStart: currentTime,
      showAlert: false,
    };
  }

  public componentDidUpdate = (prevState: IState) => {
    const { showAlert } = this.state;
    if (showAlert && !prevState.showAlert) {
      setTimeout(() => this.setState({ showAlert: false }), 5000);
    }
  }

  public render() {
    const { periodEnd, periodStart, roomEnd, roomStart, redirect, showAlert } = this.state;
    const errorPeriodEndBeforeStart =
      periodEnd <= periodStart
        ? errorObjectSeasonEndTooEarly
        : undefined;

    const errorApplicationEndBeforeStart =
      roomEnd <= roomStart
        ? errorObjectRoomEndTooEarly
        : undefined;

    const buttonDisable =
      errorPeriodEndBeforeStart !== undefined
      || errorApplicationEndBeforeStart !== undefined;

    const alertFail = showAlert
      ? this.createAlert(_ERROR_MESSAGE)
      : undefined;

    if (redirect) return (<Redirect to="/admin" />);
    return (
      <Presentational
        buttonDisable={buttonDisable}
        alertApplicationEndBeforeStart={errorApplicationEndBeforeStart}
        alertPeriodEndBeforeStart={errorPeriodEndBeforeStart}
        createFields={this.createFields}
        postApplicationSeason={this.submitApplicationSeason}
        alertFail={alertFail}
      />
    );
  }

  private submitApplicationSeason = () => {
    const { periodEnd, periodStart, roomEnd, roomStart } = this.state;
    const body = {
      newPeriodEnd: periodEnd.format(format),
      newPeriodStart: periodStart.format(format),
      newRoomEnd: roomEnd.format(format),
      newRoomStart: roomStart.format(format),
    };
    postSeason(body)
      .then(() => this.setState({ redirect: true }), () => this.setState({ showAlert: true }));
  }

  private createDateInputField = (label: string, key: string, value: Moment) => {
    return (
      <DateInputField
        key={label}
        label={label}
        value={value}
        objectKey={key}
        setDate={this.setDate}
      />
    );
  }

  private createAlert = (text: string) =>
    <AlertStripe type="advarsel" solid={true}> {text} </AlertStripe>

  private createFields = (index: number, end: number) => {
    const { periodStart, periodEnd, roomStart, roomEnd } = this.state;
    const stateEntries = Object.entries({ periodStart, periodEnd, roomStart, roomEnd });
    const fields = [];
    for (let i = index; i < end; i += 1) {
      fields.push(this.createDateInputField(
        inputTextArray[i],
        stateEntries[i][0],
        stateEntries[i][1],
      ));
    }
    return fields;
  }

  private setDate = (key: string, time: Moment) => this.setState({ [key]: time });
}

export default CreateSeason;
