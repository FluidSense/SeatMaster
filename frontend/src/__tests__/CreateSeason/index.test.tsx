import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import moment from 'moment';
import * as React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import CreateSeason, { setTime } from '../../components/CreateSeason';

const mockStore = configureMockStore();
describe('Create season', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<CreateSeason />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Creates correct alert', () => {
    const wrapper = shallow(<CreateSeason />);
    console.log(wrapper.dive().find('#new-season-btn').html());
    // const testStripe = <AlertStripe type="advarsel" solid={true}>Test</AlertStripe>;

  });
});

describe('Test const functions', () => {
  it('Tests if moment is properly set to 23:59', () => {
    const thisMoment = moment();
    const momentMidnight = setTime(thisMoment);
    expect(thisMoment.set({ hour: 23, minute: 59 })).toEqual(momentMidnight);
  });
});
