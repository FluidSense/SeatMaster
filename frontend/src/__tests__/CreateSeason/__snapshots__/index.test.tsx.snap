// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create season renders correctly 1`] = `
<Presentational
  buttonDisable={false}
  createFields={[Function]}
  postApplicationSeason={[Function]}
/>
`;
